### ISSUE TITLE
should c_ptrTo on a c_array return a pointer to the element?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#24156]
`c_ptrTo` on a Chapel array returns a pointer to the initial element.
But on a `c_array` it returns a pointer to the `c_array` itself.
Of course, such a pointer has the same value as a pointer to the first element.

Anyway, this issue asks if `c_ptrTo` / `c_ptrToConst` should be extended to have the pointer-to-first-element behavior for `c_array` to match the other Chapel array types.

In other words, it asks if this program should print out `c_ptr(int)` rather than the `c_ptr(c_array(int(64),4))` it prints now:

``` chapel
use CTypes;

proc main() {
  var carr: c_array(int, 4);

  var ptr = c_ptrTo(carr);

  writeln(ptr.type:string); // c_ptr(c_array(int(64),4))
}
```

Note that C++ does not seem to do this when using `&`, however the `c_addrOf` vs `c_ptrTo` distinction is unique to Chapel, and arguably the `c_addrOf` is the one that is intended to match `&`.

C++ Example:

<details>

``` c++
#include <typeinfo>
#include <cstdio>

int main() {
  int arr[5];
  printf("%s\n", typeid(arr).name()); // A5_i
  printf("%s\n", typeid(&arr).name()); // PA5_i
  int* ptr = arr;
  printf("%s\n", typeid(ptr).name()); // Pi
}
```

</details>

Also note that both C++ and Chapel allow implicitly converting a `c_array` to a pointer to the initial element, so this compiles in Chapel:

``` chapel
use CTypes;

proc main() {
  var carr: c_array(int, 4);

  var ptr2: c_ptr(int) = carr;
  writeln(ptr2.type:string); // c_ptr(int(64))
}

**Comment by bradcray** [comment]
I agree that `c_ptrTo()` on a `c_array` would be more intuitive and consistent if it were to point to the first element in the array, and that `c_addrOf()` makes sense to have be the thing pointing to the `c_array`.  (Does that make it effectively a pointer-to-a-pointer to the first element, just with a different type name than `c_ptr(c_ptr(int))`?)

**Comment by bradcray** [comment]
(it'd be particularly good to get @riftEmber's thoughts on this as well, given her work on these routines)

**Comment by mppf** [comment]
> Does that make it effectively a pointer-to-a-pointer to the first element, just with a different type name than `c_ptr(c_ptr(int))`?)

No, since the `c_array` itself starts in memory at the same place as the first element, it's the same value as a pointer to the first element.

**Comment by bradcray** [comment]
OK, thanks.  Re-reading, I'm seeing I missed this line:

> Of course, such a pointer has the same value as a pointer to the first element.

**Comment by riftEmber** [comment]
Yes, I definitely agree with the proposed change, analogous to what we have for Chapel arrays. There's some discussion to be had on how many Chapel array conveniences we need `c_array` to have, but as `c_ptrTo`/`c_addrOf` are primarily for C interop I think this is an easy one to support.
